File: libpinger/pingtracker.py
import logging


class PingTracker:
    def __init__(self):
        self.next_sequence_nr = None
        self.sequence_nrs = []
        self.latencies = []

    def track(self, sequence_nr, latency):
        self.sequence_nrs.append(sequence_nr)
        self.latencies.append(latency)

    def calculate(self):
        def calculate_latencies():
            if not self.latencies:
                return None
            logging.debug(f'Latencies: {self.latencies}')
            return round(sum(self.latencies) / len(self.latencies), 1)

        def calculate_packet_loss():
            def process_range(series):
                gap = 0
                logging.debug(series)
                # calculate the gap between each (ordered) packet
                gaps = [series[i+1]-series[i]-1 for i in range(len(series)-1)]
                logging.debug(f'gaps: {gaps}')
                gap += sum(gaps)
                # any packets lost between now and the previous batch?
                if self.next_sequence_nr is not None:
                    gap += series[0] - self.next_sequence_nr
                logging.debug(f'Final gap: {gap}')
                return gap
            if not self.sequence_nrs:
                return None
            # sort the sequence nrs and remove all duplicates
            packets = sorted(set(self.sequence_nrs))
            logging.debug(f'Sequence Nrs received: {packets}')
            # if it's the first call, safe to assume the smallest nr is next expected
            if self.next_sequence_nr is None:
                self.next_sequence_nr = packets[0]
            # sequence numbers can wrap around!
            # In this case, we'd get something like [ 0, 1, 2, 3, 65534, 65535 ]
            # split into two series [ 65534, 65535 ] and [ 0, 1, 2 ] using next_sequence_nr as a boundary
            # process the larger range first (pre-wrap) and then the smaller one (post-wrap)
            loss = 0
            larger = list(filter(lambda i: i >= self.next_sequence_nr, packets))
            smaller = list(filter(lambda i: i < self.next_sequence_nr, packets))
            if larger:
                loss += process_range(larger)
                self.next_sequence_nr = larger[-1] + 1
            if smaller:
                self.next_sequence_nr = 0
                loss += process_range(smaller)
                self.next_sequence_nr = smaller[-1] + 1
            return loss
        latency = calculate_latencies()
        packet_loss = calculate_packet_loss()
        # set up next cycle
        self.sequence_nrs = []
        self.latencies = []
        return packet_loss, latency

File: pingtracker/pingtracker.go
package pingtracker

import (
	"sort"
	"sync"
	"time"
)

// PingTracker handle containing all required data
type PingTracker struct {
	NextSeqNr int
	seqNrs    []int
	latencies []time.Duration
	lock      sync.Mutex
}

// New creates a new PingTracker
func New() *PingTracker {
	return &PingTracker{}
}

// Track measured sequence number and latency
func (tracker *PingTracker) Track(SeqNr int, Latency time.Duration) {
	tracker.lock.Lock()
	defer tracker.lock.Unlock()

	tracker.seqNrs = append(tracker.seqNrs, SeqNr)
	tracker.latencies = append(tracker.latencies, Latency)
}

// Calculate packet loss and latency for all data stored in PingTracker
func (tracker *PingTracker) Calculate() (int, int, time.Duration) {
	tracker.lock.Lock()
	defer tracker.lock.Unlock()

	loss := tracker.calculateLoss()
	count := len(tracker.latencies)
	latency := tracker.calculateLatency()

	// empty the slice but keep the memory
	tracker.seqNrs = tracker.seqNrs[:0]
	tracker.latencies = tracker.latencies[:0]

	return count, loss, latency
}

func (tracker *PingTracker) calculateLatency() (total time.Duration) {
	count := len(tracker.latencies)
	if count == 0 {
		return
	}

	for _, entry := range tracker.latencies {
		total += entry
	}
	return
}

func (tracker *PingTracker) calculateLoss() (gap int) {
	if len(tracker.seqNrs) == 0 {
		return 0
	}
	// Sort all sequence numbers and remove duplicates
	tracker.seqNrs = unique(tracker.seqNrs)

	// sequence numbers can roll over!
	// In this case, we'd get something like [ 0, 1, 2, 3, 65534, 65535 ]
	// Split into two lists [ 65534, 65535 ] and [ 0, 1, 2 ] using nextSeqNr as a boundary
	// Process the higher list first (pre-rollover) and then the lower one (post-rollover)

	index := 0
	for index < len(tracker.seqNrs) && tracker.seqNrs[index] < tracker.NextSeqNr-60 {
		index++
	}

	// pre-rollover / no rollover
	gap = tracker.processRange(tracker.seqNrs[index:])

	if index > 0 {
		tracker.NextSeqNr = 0
		gap += tracker.processRange(tracker.seqNrs[:index])
	}

	return
}

func unique(seqNrs []int) (result []int) {
	uniqueSeqNrs := make(map[int]struct{})
	for _, seqNr := range seqNrs {
		if _, ok := uniqueSeqNrs[seqNr]; ok == false {
			uniqueSeqNrs[seqNr] = struct{}{}
			result = append(result, seqNr)
		}
	}
	sort.Ints(result)
	return
}

func (tracker *PingTracker) processRange(sequence []int) (gap int) {
	count := len(sequence)
	if count == 0 {
		panic("processRange: sequence range should not be empty")
	}

	index := 0
	// skip older packets
	for ; index < count && sequence[index] < tracker.NextSeqNr; index++ {
	}

	for ; index < count; index++ {
		gap += sequence[index] - tracker.NextSeqNr
		tracker.NextSeqNr = sequence[index] + 1
	}

	return
}
